### PR TITLE
Reorganize test run in automation

### PR DESCRIPTION
**Pull request by troydai**
This pull request is the first step towards consolidating automation scripts. Changes include:
1. Use nosetests to execute the tests in automation. Nosetests can run tests in parallel. It also generates xunit like test report in XML, which can be feed into other reporting systems.
2. Consolidate automation related to into automation package and remove duplicate and out-of-date code.
3. Clean up and add comments.
4. Add basics code coverage facility. It now generates code coverage data. The accuracy of the results needs further study.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -2,6 +2,7 @@ adal==0.4.3
 applicationinsights==0.10.0
 argcomplete==1.3.0
 colorama==0.3.7
+coverage==4.2
```
Comment (tjprescott) [inline requirements.txt]: Does this file only apply to dev-setup or all installation methods? It would seem that if someone did a curl install of the CLI for production use, then they wouldn't need (or probably want) packages related strictly to testing and development.

File: scripts/automation/coverage/run.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+from coverage import Coverage
+
+import azure.cli.core.application as cli_application
+from ..tests.nose_helper import get_nose_runner
+from ..utilities.path import get_core_modules_paths_with_tests, \
+    get_command_modules_paths_with_tests, get_repo_root, get_test_results_dir, make_dirs
+
+
+# TODO: Fix track command logic in vcr_test_base.py.
```
Comment (tjprescott) [inline scripts/automation/coverage/run.py]: I'm assuming this is the traditional line or branch based code coverage metric, or is it a port of the command coverage logic (which needs to be augmented, but that's a separate thing).

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py [hunk 3]
```diff
@@ -1064,6 +1064,9 @@ def body(self):
         user_name = 'myadmin'
         config_file = _write_config_file(user_name)
 
+        if not os.path.exists(config_file):
```
Comment (tjprescott) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py]: Is this related to the PR?

File: requirements.txt (hunk 1 above)
Comment (troydai) [inline requirements.txt]: This is for test only. Where do we list the requirements for dev environment?

File: scripts/automation/coverage/run.py (hunk 2 above)
Comment (troydai) [inline scripts/automation/coverage/run.py]: This is the code I ported from the command coverage logic. In my test, the command coverage doesn't work. I may not have the full context so I add this comment for following up.

File: src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py (hunk 3 above)
Comment (troydai) [inline src/command_modules/azure-cli-vm/azure/cli/command_modules/vm/tests/test_vm_commands.py]: Good catch. This is not related. I'll remove it.

File: requirements.txt (hunk 1 above)
Comment (tjprescott) [inline requirements.txt]: There's a `requirements.txt` file in the CLI's root directory.

Comment (troydai) [inline requirements.txt]: And I don't think this `requirement.txt` should impact the user dependencies which are defined in the `setup.py` as far as I know. @derekbekoe

File: scripts/automation/coverage/run.py (hunk 2 above)
Comment (tjprescott) [inline scripts/automation/coverage/run.py]: So the code is in place but not actively being used right now?

Comment (troydai) [inline scripts/automation/coverage/run.py]: You're right. It is not actively being used.

File: scripts/automation/coverage/run.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    print('Total commands {}, Tested commands {}, Untested commands {}'.format(
+        len(cmd_set),
+        len(commands_tested_set),
+        len(cmd_set) - len(commands_tested_set)))
+    print('COMMAND COVERAGE {0:.2f}%'.format(percentage_tested))
+
+
+class CoverageContext(object):
+    def __init__(self):
+        self._cov = Coverage(cover_pylib=False)#data_suffix='azure', cover_pylib=False)
```
Comment (derekbekoe) [inline scripts/automation/coverage/run.py]: delete the comment?

Comment (troydai) [inline scripts/automation/coverage/run.py]: will do.

File: scripts/automation/coverage/run.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    test_modules = itertools.chain(get_core_modules_paths_with_tests(),
+                                   get_command_modules_paths_with_tests())
+
+    # run code coverage on each project
+    for index, (name, _, test_path) in enumerate(test_modules):
+        with CoverageContext():
+            run_nose(name, test_path)
+
+        import shutil
+        shutil.move('.coverage', os.path.join(test_results_folder, '.covarege.{}'.format(name)))
```
Comment (troydai) [inline scripts/automation/coverage/run.py]: It's a type. I'll fix it.
